File: autokeren/tools/git.py
"""Git helpers."""
from __future__ import annotations

import subprocess
from pathlib import Path

from autokeren.tools.base import Tool, ToolResult


class _GitBase(Tool):
    """Mixin dasar untuk semua tool git."""

    def __init__(self, project_root: Path) -> None:
        self.project_root = project_root

    def _git(self, args: list[str]) -> ToolResult:
        try:
            result = subprocess.run(
                ["git", "-C", str(self.project_root)] + args,
                capture_output=True,
                text=True,
                timeout=30,
            )
            return ToolResult(output=result.stdout, error=result.stderr or None, ok=result.returncode == 0)
        except Exception as e:
            return ToolResult(error=f"{type(e).__name__}: {e}", ok=False)


class GitStatusTool(_GitBase):
    name = "git_status"
    description = "Show git status for the project root."
    parameters = {"type": "object", "properties": {}, "required": []}

    def run(self, **_: object) -> ToolResult:
        return self._git(["status", "--short"])


class GitDiffTool(_GitBase):
    name = "git_diff"
    description = "Show git diff for current changes. Pass staged=true to see staged changes only."
    parameters = {
        "type": "object",
        "properties": {
            "staged": {"type": "boolean", "description": "If true, show only staged (--cached) diff."},
        },
        "required": [],
    }

    def run(self, staged: bool = False, **_: object) -> ToolResult:
        args = ["diff", "--cached"] if staged else ["diff"]
        return self._git(args)


class GitCommitTool(_GitBase):
    name = "git_commit"
    description = "Stage all changes and commit with the given message following Conventional Commits format."
    requires_permission = True
    parameters = {
        "type": "object",
        "properties": {
            "message": {"type": "string", "description": "Commit message (use Conventional Commits: feat:, fix:, docs:, etc.)"},
        },
        "required": ["message"],
    }

    def run(self, message: str, **_: object) -> ToolResult:
        self._git(["add", "-A"])
        return self._git(["commit", "-m", message])

    def permission_desc(self, message: str = "", **_: object) -> str:
        return f"git commit: {message[:60]}"


class GitLogTool(_GitBase):
    name = "git_log"
    description = "Show recent git commit history with author, date, and message."
    parameters = {
        "type": "object",
        "properties": {
            "n": {"type": "integer", "description": "Number of recent commits to show (default: 10)."},
            "oneline": {"type": "boolean", "description": "If true, show compact one-line format."},
        },
        "required": [],
    }

    def run(self, n: int = 10, oneline: bool = False, **_: object) -> ToolResult:
        fmt = ["--oneline"] if oneline else ["--pretty=format:%h  %an  %ar  %s"]
        return self._git(["log", f"-{n}"] + fmt)


class GitBranchTool(_GitBase):
    name = "git_branch"
    description = "List, create, or switch git branches."
    requires_permission = True
    parameters = {
        "type": "object",
        "properties": {
            "action": {
                "type": "string",
                "enum": ["list", "create", "switch", "create_and_switch"],
                "description": "Action to perform on branches.",
            },
            "name": {"type": "string", "description": "Branch name (required for create/switch)."},
        },
        "required": ["action"],
    }

    def run(self, action: str, name: str = "", **_: object) -> ToolResult:
        if action == "list":
            return self._git(["branch", "-a"])
        if action == "create":
            if not name:
                return ToolResult(error="Branch name required.", ok=False)
            return self._git(["branch", name])
        if action == "switch":
            if not name:
                return ToolResult(error="Branch name required.", ok=False)
            return self._git(["checkout", name])
        if action == "create_and_switch":
            if not name:
                return ToolResult(error="Branch name required.", ok=False)
            return self._git(["checkout", "-b", name])
        return ToolResult(error=f"Unknown action: {action}", ok=False)

    def permission_desc(self, action: str = "", name: str = "", **_: object) -> str:
        return f"git branch {action} {name}".strip()

    def needs_permission(self, action: str = "", **_: object) -> bool:
        return action != "list"


class GitAutoCommitTool(_GitBase):
    name = "git_auto_commit"
    description = "Auto-stage and commit modified files with an auto-generated Conventional Commits message."
    requires_permission = False
    parameters = {
        "type": "object",
        "properties": {
            "files": {
                "type": "array",
                "items": {"type": "string"},
                "description": "List of modified file paths to stage and commit.",
            },
            "summary": {"type": "string", "description": "Short task summary used to build the commit message."},
        },
        "required": ["files", "summary"],
    }

    def run(self, files: list[str], summary: str, **_: object) -> ToolResult:
        if not files:
            return ToolResult(output="Tidak ada file dimodifikasi, skip commit.", ok=True)
        check = self._git(["rev-parse", "--is-inside-work-tree"])
        if not check.ok:
            return ToolResult(error="Bukan git repository, auto-commit dilewati.", ok=False)
        has_changes = self._git(["status", "--porcelain"])
        if not has_changes.ok or not str(has_changes.output or "").strip():
            return ToolResult(output="Tidak ada perubahan untuk di-commit.", ok=True)
        prefix = self._infer_prefix(files)
        scope = self._infer_scope(files)
        scope_str = f"({scope})" if scope else ""
        msg = f"{prefix}{scope_str}: {summary[:72]}"
        body_lines = [f"- {f}" for f in files[:20]]
        full_msg = msg + "\n\n" + "\n".join(body_lines)
        stage = self._git(["add"] + [str(f) for f in files])
        if not stage.ok:
            return stage
        return self._git(["commit", "-m", full_msg])

    def _infer_prefix(self, files: list[str]) -> str:
        names = " ".join(files).lower()
        if "test" in names:
            return "test"
        if any(k in names for k in ("readme", ".md", "docs", "changelog")):
            return "docs"
        if any(k in names for k in ("config", "yaml", "toml", "pyproject")):
            return "chore"
        return "feat"

    def _infer_scope(self, files: list[str]) -> str:
        parts: list[str] = []
        for f in files:
            p = Path(f)
            if len(p.parts) > 1:
                parts.append(p.parts[-2])
        unique = list(dict.fromkeys(parts))
        return ",".join(unique[:2])

File: internal/tool/git.go
package tool

import (
	"context"
	"fmt"
	"os/exec"
	"strings"
)

type GitStatus struct{ Root string }

func (t GitStatus) Definition() Definition {
	return Definition{Name: "git_status", Description: "Show git working tree status.", Parameters: map[string]any{"type": "object"}}
}
func (t GitStatus) NeedsPermission(map[string]any) (bool, string) { return false, "" }
func (t GitStatus) Run(ctx context.Context, _ map[string]any, _ Emitter) Result {
	return gitRun(ctx, t.Root, "status", "--short")
}

type GitDiff struct{ Root string }

func (t GitDiff) Definition() Definition {
	return Definition{Name: "git_diff", Description: "Show git diff.", Parameters: map[string]any{"type": "object", "properties": map[string]any{"staged": map[string]any{"type": "boolean"}}}}
}
func (t GitDiff) NeedsPermission(map[string]any) (bool, string) { return false, "" }
func (t GitDiff) Run(ctx context.Context, args map[string]any, _ Emitter) Result {
	if staged, ok := args["staged"].(bool); ok && staged {
		return gitRun(ctx, t.Root, "diff", "--cached")
	}
	return gitRun(ctx, t.Root, "diff")
}

type GitLog struct{ Root string }

func (t GitLog) Definition() Definition {
	return Definition{Name: "git_log", Description: "Show recent git commits.", Parameters: map[string]any{"type": "object"}}
}
func (t GitLog) NeedsPermission(map[string]any) (bool, string) { return false, "" }
func (t GitLog) Run(ctx context.Context, _ map[string]any, _ Emitter) Result {
	return gitRun(ctx, t.Root, "log", "-10", "--oneline", "--decorate")
}

type GitCommit struct{ Root string }

func (t GitCommit) Definition() Definition {
	return Definition{Name: "git_commit", Description: "Create a git commit.", RequiresPermission: true, Parameters: map[string]any{"type": "object", "properties": map[string]any{"message": map[string]any{"type": "string"}}, "required": []string{"message"}}}
}
func (t GitCommit) NeedsPermission(args map[string]any) (bool, string) {
	m, _ := args["message"].(string)
	return true, "Create git commit: " + m
}
func (t GitCommit) Run(ctx context.Context, args map[string]any, _ Emitter) Result {
	m, _ := args["message"].(string)
	if strings.TrimSpace(m) == "" {
		return Result{OK: false, Error: "message is required"}
	}
	return gitRun(ctx, t.Root, "commit", "-am", m)
}

type GitBranch struct{ Root string }

func (t GitBranch) Definition() Definition {
	return Definition{Name: "git_branch", Description: "List, create, or switch git branches.", RequiresPermission: true, Parameters: map[string]any{"type": "object", "properties": map[string]any{"action": map[string]any{"type": "string", "enum": []string{"list", "create", "switch", "create_and_switch"}}, "name": map[string]any{"type": "string"}}, "required": []string{"action"}}}
}
func (t GitBranch) NeedsPermission(args map[string]any) (bool, string) {
	a, _ := args["action"].(string)
	return a != "list", "Change git branch"
}
func (t GitBranch) Run(ctx context.Context, args map[string]any, _ Emitter) Result {
	a, _ := args["action"].(string)
	name, _ := args["name"].(string)
	switch a {
	case "list":
		return gitRun(ctx, t.Root, "branch", "-a")
	case "create":
		if name == "" {
			return Result{OK: false, Error: "branch name is required"}
		}
		return gitRun(ctx, t.Root, "branch", name)
	case "switch":
		if name == "" {
			return Result{OK: false, Error: "branch name is required"}
		}
		return gitRun(ctx, t.Root, "switch", name)
	case "create_and_switch":
		if name == "" {
			return Result{OK: false, Error: "branch name is required"}
		}
		return gitRun(ctx, t.Root, "switch", "-c", name)
	default:
		return Result{OK: false, Error: "unknown branch action"}
	}
}

type GitAutoCommit struct{ Root string }

func (t GitAutoCommit) Definition() Definition {
	return Definition{Name: "git_auto_commit", Description: "Stage selected files and create a conventional commit.", RequiresPermission: true, Parameters: map[string]any{"type": "object", "properties": map[string]any{"files": map[string]any{"type": "array"}, "summary": map[string]any{"type": "string"}}, "required": []string{"files", "summary"}}}
}
func (t GitAutoCommit) NeedsPermission(map[string]any) (bool, string) {
	return true, "Create automatic git commit"
}
func (t GitAutoCommit) Run(ctx context.Context, args map[string]any, _ Emitter) Result {
	files, _ := args["files"].([]any)
	summary, _ := args["summary"].(string)
	if strings.TrimSpace(summary) == "" {
		return Result{OK: false, Error: "summary is required"}
	}
	if len(files) == 0 {
		return Result{OK: true, Output: "Tidak ada file untuk di-commit."}
	}
	paths := make([]string, 0, len(files))
	for _, raw := range files {
		if path, ok := raw.(string); ok && path != "" {
			paths = append(paths, path)
		}
	}
	if len(paths) == 0 {
		return Result{OK: true, Output: "Tidak ada file untuk di-commit."}
	}
	addArgs := append([]string{"add", "--"}, paths...)
	if result := gitRun(ctx, t.Root, addArgs...); !result.OK {
		return result
	}
	return gitRun(ctx, t.Root, "commit", "-m", "feat: "+summary)
}
func gitRun(ctx context.Context, root string, args ...string) Result {
	cmd := exec.CommandContext(ctx, "git", args...)
	cmd.Dir = root
	out, err := cmd.CombinedOutput()
	if err != nil {
		return Result{OK: false, Output: string(out), Error: err.Error()}
	}
	return Result{OK: true, Output: string(out)}
}

var _ = fmt.Sprintf
